corehq/pillows/synclog.py

```diff
@@ -0,0 +1,139 @@
+
+from casexml.apps.phone.models import SyncLog
+from casexml.apps.phone.dbaccessors.sync_logs_by_user import get_synclogs_for_user
+
+
+def get_synclog_pillow(pillow_id='SynclogPillow'):
+    """
+    This gets a pillow which iterates through all synclogs
+    """
+    couch_db = SyncLog.get_db()
+    change_feed = CouchChangeFeed(couch_db, include_docs=True)
+    checkpoint = PillowCheckpoint('synclog', couch_db)
+    form_processor = SynclogProcessor()
+    return ConstructedPillow(
+        name=pillow_id,
+        checkpoint=checkpoint,
+        change_feed=change_feed,
+        processor=form_processor,
+        change_processed_event_handler=PillowCheckpointEventHandler(
+            checkpoint=checkpoint, checkpoint_frequency=100
+        ),
+    )
+
+
+class SynclogProcessor(PillowProcessor):
+
+    def process_change(self, pillow_instance, change):
+
+        synclog = change.get_document()
+        if not synclog:
+            return
+
+        version = None
+        app_id = None
+        try:
+            last_sync_date = string_to_utc_datetime(synclog.get('date'))
+        except ValueError:
+            return
+        build_id = synclog.get('build_id')
+        if build_id:
+            version, app_id = get_version_and_app_from_build_id(synclog.get('domain'), build_id)
+        user_id = synclog.get('user_id')
+
+        if user_id:
+            user = CouchUser.get_by_user_id(user_id)
+
+            last_sync = filter_by_app(user.reporting_metadata.last_syncs, app_id)
+
+            if last_sync is None or last_sync_date >= last_sync.sync_date:
+                if last_sync is None:
+                    last_sync = LastSync()
+                    user.reporting_metadata.last_syncs.append(last_sync)
+                last_sync.sync_date = last_sync_date
+                last_sync.build_version = version
+                last_sync.app_id = app_id
+
+                if user.reporting_metadata.last_sync_for_user is None \
+                        or last_sync_date > user.reporting_metadata.last_sync_for_user.sync_date:
+                    user.reporting_metadata.last_sync_for_user = last_sync
+
+                if version:
+                    update_latest_builds(user, app_id, last_sync_date, version)
+
+                user.save()
+
+
+class SynclogReindexerDocProcessor(BaseDocProcessor):
+
+    def __init__(self, pillow_processor):
+        self.pillow_processor = pillow_processor
+
+    def process_doc(self, doc):
+        synclog_changes = self._doc_to_changes(doc)
+        for change in synclog_changes:
+            try:
+                self.pillow_processor.process_change(None, change)
+            except Exception:
+                return False
+        return True
+
+    def handle_skip(self, doc):
+        print('Unable to process user {}'.format(
+            doc['_id'],
+        ))
+        return True
+
+    def _doc_to_changes(self, doc):
+        synclogs = get_synclogs_for_user(doc['_id'])
```

Comment:
I know that this function has a default of 10 for limit but adding in here explicitly would make it much clearer that we're intentionally only fetching the last 10. As it stands it looks like we're fetching all the synclogs for the user.